### DISCUSSION TITLE
Simple console split Spectre/System.Commandline

### DISCUSSION THREAD

**Discussion by oo777** [spectreconsole/spectre.console#1585] [General]
I tried multiple approaches without success but the requirement is pretty simple:

- Using Spectre to print out a status table
- Using System.Commandline to enter user commands
- Using both together for long running tasks

The status table I used the "Live" option and it's fine and flicker free but I need a prompt under the table handling inputs form another non spectre task passing it to System.Commandline handlers resulting in an updated Spectre result table above the prompt.

Any ideas here?

**Comment by nils-a** [comment]
Have you looked at [GUI.cs](https://github.com/gui-cs) ?

**Reply by oo777** [reply]
Thanks! I gave it a look, tried spectre with "AlternateScreen" afterwards without success. Now I switched to gui-cs like this:
using System.CommandLine;
using System.Text;
using Terminal.Gui;
// ReSharper disable ConvertToPrimaryConstructor

namespace MPMApp;

public static class ConsoleGui {

    public static async Task Initialize( RootCommand? rootCommand ) {
        
        _rootCommand = rootCommand;

        Application.Init();
        var top = Application.Top;

        // Create a window
        var win = new Window( "Console Redirection Example" ) {
            X = 0,
            Y = 0,
            Width = Dim.Fill(),
            Height = Dim.Fill()
        };
        top.Add( win );

        // Create a TextView for stdout
        _outputView = new TextView() {
            X = 0,
            Y = 0,
            Width = Dim.Fill(),
            Height = Dim.Percent( 80 ),
            ReadOnly = true
        };
        win.Add( _outputView );

        // Create a TextField for stdin
        _inputField = new TextField {
            X = 0,
            Y = Pos.Bottom( _outputView ),
            Width = Dim.Fill(),
            Height = 1
        };
        win.Add( _inputField );

        // Redirect stdout to the TextView
        var textViewWriter = new TextViewWriter( _outputView );
        Console.SetOut( textViewWriter );

        // Display the initial prompt
        DisplayPrompt();

        // Handle input from the TextField
        _ = HandleInputAsync();

        // Run the Terminal.Gui application
        Application.Run();
    }

    public static void WriteLine( string message ) {
        Console.WriteLine( message );
    }

    public static int GetCommandResult() {
        return _commandResult;
    }


    /***********************************************************
     *
     *   PRIVATE
     *
     **********************************************************/

    private static void DisplayPrompt() {
        Application.MainLoop.Invoke( () => {
            if( _inputField != null ) {
                _inputField.Text = _prompt;
                _inputField.CursorPosition = _prompt.Length;
            }
        } );
    }


    private static async Task HandleInputAsync() {
        while( true ) {
            await Task.Delay( 100 ); // Polling interval

            Application.MainLoop.Invoke( () => {
                if( _inputField.HasFocus ) {
                    Application.Top.KeyPress += ( keyEvent ) => {
                        if( keyEvent.KeyEvent.Key == Key.Enter ) {
                            var input = _inputField.Text.ToString().TrimStart( '>' ); // Trim the prompt
                            _inputField.Text = string.Empty;

                            // Process input (for example, echo it back)
                            WriteLine( _prompt + input );

                            // Invoke the command asynchronously
                            _rootCommand.InvokeAsync( input );
                            //_ = Task.Run( async () => await _rootCommand.InvokeAsync( input ) );

                            // Display the prompt again
                            DisplayPrompt();
                        }
                    };
                }
            } );
        }
    }


    private static TextView? _outputView;
    private static TextField? _inputField;
    private static RootCommand? _rootCommand;
    private static readonly string _prompt = "> ";
    private static int _commandResult;
}


public class TextViewWriter : TextWriter {

    public TextViewWriter( TextView? outputView ) {
        _outputView = outputView;
        _stringWriter = new StringWriter();
    }

    public override Encoding Encoding {
        get => Encoding.UTF8;
    }

    public override void Write( char value ) {
        _stringWriter.Write( value );
        Application.MainLoop.Invoke( () => {
            if( _outputView != null ) {
                _outputView.Text = _stringWriter.ToString();
                _outputView.ScrollTo( _outputView.Lines - 1 );
            }
        } );
    }

    public override void Write( string? value ) {
        _stringWriter.Write( value );
        Application.MainLoop.Invoke( () => {
            if( _outputView != null ) {
                _outputView.Text = _stringWriter.ToString();
                _outputView.ScrollTo( _outputView.Lines - 1 );
            }
        } );
    }

    public override void WriteLine( string? value ) {
        _stringWriter.WriteLine( value );
        Application.MainLoop.Invoke( () => {
            if( _outputView != null ) {
                _outputView.Text = _stringWriter.ToString();
                _outputView.ScrollTo( _outputView.Lines - 1 );
                _outputView.SetNeedsDisplay();
            }
        } );
    }

    /***********************************************************
     *
     *   PRIVATE
     *
     **********************************************************/
    private readonly TextView? _outputView;
    private readonly StringWriter _stringWriter;
}
